blog: Windows Operational Readiness (#45130)
kubernetes/website: content/en/blog/_posts/2024-windows-ops-readiness.md

---
layout: blog
title: "Introducing the Windows Operational Readiness Specification"
date: TBA
slug: TBA

> canonicalUrl: TBA

Reviewer: ```suggestion
```
(not needed unless the original article appears on a different Kubernetes blog)

Some notable examples brought to the attention of SIG Windows were:

- An issue with [loadBalancerSourceRanges](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.29/#:~:text=loadBalancerSourceRanges) functionality not operating correctly on Windows nodes, detailed in a GitHub issue [kubernetes/kubernetes#120033](https://github.com/kubernetes/kubernetes/issues/120033).
- Reports of functionality issues with Windows features, such as “[GMSA](https://learn.microsoft.com/en-us/windows-server/security/group-managed-service-accounts/group-managed-service-accounts-overview) isn’t working, why?”, discussed in [microsoft/Windows-Containers#44](https://github.com/microsoft/Windows-Containers/issues/44).
- Challenges developing networking policy tests that could objectively evaluate Container Network Interfaces (CNIs) across different operating system configurations, as discussed in [kubernetes/kubernetes#97751](https://github.com/kubernetes/kubernetes/issues/97751)

> The SIG Windows community therefore recognized the need for a tailored solution to ensure Windows nodes' operational readiness *before* their deployment into production environments. Thus, the Windows Operational Readiness Specification was born.

Reviewer 2: ```suggestion
The SIG Windows community therefore recognized the need for a tailored solution to ensure Windows nodes' operational readiness *before* their deployment into production environments. Thus, the [Windows Operational Readiness Specification](https://github.com/kubernetes/enhancements/blob/d60c6149551b7f852c16842fb952a499c2cc4315/keps/sig-windows/2578-windows-conformance/README.md) was born.
```

In the Linux world, tools such as [Sonobuoy](https://sonobuoy.io/) simplify execution of the conformance suite, relieving users from needing to be aware of Kubernetes' compilation paths or the semantics of Ginkgo tags.

Regarding needing to compile the Kubernetes tests, we realized that Windows users might similarly find the process of compiling and running the Kubernetes e2e suite from scratch similarly undesirable, hence, there was a clear need to provide a user-friendly, "push-button" solution that is ready to go. Moreover, regarding Ginkgo tags, applying conformance tests to Windows nodes through a set of [Ginkgo](https://onsi.github.io/ginkgo/) tags would also be burdensome for any user, including Linux enthusiasts or experienced Windows system admins alike.

To bridge the gap and give users a straightforward way to confirm their clusters support a variety of features, the SIG Windows community found it necessary to therefore create the Windows Operational Readiness application. This application written in Go, simplifies the process to run the necessary Windows specific tests while delivering results in a clear, accessible format.

> This initiative has been a collaborative effort, with contributions from different cloud providers and platforms including AWS, Azure and VMware.

Reviewer 2: ```suggestion
This initiative has been a collaborative effort, with contributions from different cloud providers and platforms, including Amazon, Microsoft, SUSE, and Broadcom.
```

| ------------------------ | ------------------------------------------------------------ |
| `Core.Network`           | Tests minimal networking functionality, e.g. ability to access pod-by-pod IP. |
| `Core.Storage`           | Tests minimal storage functionality, e.g. ability to mount a hostPath storage volume. |
| `Core.Scheduling`        | Tests minimal scheduling functionality, e.g. ability to schedule a pod with CPU limits. |
| `Core.Concurrent`        | Tests minimal concurrent functionality, e.g. the ability of a node to handle traffic to multiple pods concurrently. |
| `Extend.HostProcess`     | Tests features related to Windows HostProcess pod functionality. |
| `Extend.ActiveDirectory` | Tests features related to Active Directory functionality.    |
| `Extend.NetworkPolicy`   | Tests features related to Network Policy functionality.      |
| `Extend.Network`         | Tests advanced networking functionality, e.g. ability to support IPv6 |
| `Extend.Worker`          | Tests features related to Windows worker node functionality, e.g. ability for nodes to access TCP and UDP services in the same cluster |

# How to Conduct Operational Readiness Tests for Windows Nodes

> To run the Windows Operational Readiness test suite, refer to the project’s [README.md](https://github.com/kubernetes-sigs/windows-operational-readiness/blob/main/README.md), which provides more elaborate instructions. The project offers flexibility in how you can execute tests, and you can run the tests using a compiled binary or a Sonobuoy plugin. You also have the choice to run the tests against the entire test suite or by specifying a list of categories. Cloud providers may execute these tests and upload their conformance results, enhancing transparency and reliability.

Reviewer 3: ```suggestion
To run the Windows Operational Readiness test suite, refer to the project’s [README.md](https://github.com/kubernetes-sigs/windows-operational-readiness/blob/main/README.md), which provides more elaborate instructions. The project offers flexibility in how you can execute tests, either using a compiled binary or a Sonobuoy plugin. You also have the choice to run the tests against the entire test suite or by specifying a list of categories. Cloud providers have the choice of uploading their conformance results, enhancing transparency and reliability.
```

```

# Looking Ahead

We’re looking to improve our curated list of Windows-specific tests by adding new tests to the Kubernetes repository and also identifying existing test cases that can be targetted. The long term goal for this subproject is to continually enhance test coverage for Windows worker nodes and improve the robustness of Windows support, facilitating a seamless experience across diverse cloud environments. We also have plans to integrate the Windows Operational Readiness tests into the official Kubernetes conformance suite.

If you are interested in helping us out, please reach out to us! We welcome help in any form, from giving once-off feedback to making a code contribution, to having long-term owners to help us drive changes. The Windows Operational Readiness is owned by the SIG Windows team as a sub-project. You can reach out to the team on the [Kubernetes Slack workspace](https://communityinviter.com/apps/kubernetes/community) **#sig-windows** channel. You can also explore the [Windows Operational Readiness GitHub project](https://raw.githubusercontent.com/kubernetes-sigs/windows-operational-readiness) and make contributions directly to the project.

> Special thanks to Kulwant Singh (AWS), Pramita Gautam Rana (VMWare), Xinqi Li (Google) for their help in making notable contributions to the project. Additionally, appreciation goes to James Sturtevant (Microsoft), Mark Rossetti (Microsoft), Claudiu Belu (Cloudbase Solutions) and Aravindh Puthiyaparambil (Red Hat) from the SIG Windows team for being supportive of the project by providing guidance and helping overcome various hurdles during the initial phases of the subproject.

Reviewer 3: ```suggestion
Special thanks to Kulwant Singh (AWS), Pramita Gautam Rana (VMWare), Xinqi Li (Google) for their help in making notable contributions to the project. Additionally, appreciation goes to James Sturtevant (Microsoft), Mark Rossetti (Microsoft), Claudiu Belu (Cloudbase Solutions) and Aravindh Puthiyaparambil (Softdrive Technologies Group Inc.) from the SIG Windows team for being supportive of the project by providing guidance and helping overcome various hurdles during the initial phases of the subproject.
```
